core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1647,69 +1656,51 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
-  def handleJoinGroupRequest(request: RequestChannel.Request, requestLocal: RequestLocal): Unit = {
-    val joinGroupRequest = request.body[JoinGroupRequest]
+  private def makeGroupCoordinatorRequestContext(
+    request: RequestChannel.Request,
+    requestLocal: RequestLocal
+  ): GroupCoordinatorRequestContext = {
+    new GroupCoordinatorRequestContext(
+      request.context.header.data.requestApiVersion,
+      request.context.header.data.clientId,
+      request.context.clientAddress,
+      requestLocal.bufferSupplier
+    )
+  }
 
-    // the callback for sending a join-group response
-    def sendResponseCallback(joinResult: JoinGroupResult): Unit = {
-      def createResponse(requestThrottleMs: Int): AbstractResponse = {
-        val responseBody = new JoinGroupResponse(
-          new JoinGroupResponseData()
-            .setThrottleTimeMs(requestThrottleMs)
-            .setErrorCode(joinResult.error.code)
-            .setGenerationId(joinResult.generationId)
-            .setProtocolType(joinResult.protocolType.orNull)
-            .setProtocolName(joinResult.protocolName.orNull)
-            .setLeader(joinResult.leaderId)
-            .setSkipAssignment(joinResult.skipAssignment)
-            .setMemberId(joinResult.memberId)
-            .setMembers(joinResult.members.asJava),
-          request.context.apiVersion
-        )
+  def handleJoinGroupRequest(
+    request: RequestChannel.Request,
+    requestLocal: RequestLocal
+  ): CompletableFuture[Unit] = {
```

Comment:
some questions for my understanding:
1. it looks like we handle exceptions thrown by the new joinGroup in `newGroupCoordinator.joinGroup(ctx, joinGroupRequest.data).handle[Unit]`. are we returning a future here to handle exceptions thrown during `sendResponse()`? it looks to me that would be handled by `case e: Throwable => handleError(e)` or am i missing something?
3. would `GroupCoordinatorAdapter.joinGroup()` work without returning a CompletableFuture object? 
4. are we returning CompletableFuture in GroupCoordinatorAdapter to prepare it for the new group coordinator since it will use multiple threads to handle a single join/sync group request?

Reply:
1. Yeah, that's right. I return a future to catch any other errors. It could be from `sendResponse` but it could also be other issues.
2. No, it would not because the underlying API uses a callback.
3. All APIs will be async for this reason but not that most of them are already async today. The difference is that we use a future here instead of a callback.

Reply:
Sorry If i missed it, but what was the reasoning of using futures instead of callbacks?

Reply:
i am also curious, that is one of my confusions :sweat_smile: 

> All APIs will be async for this reason but not that most of them are already async today

can you elaborate on the reason and clarify whether most of the APIs are async or not? i was under the impression that a single API request -> response is handled synchronously by a request handler thread.

Reply:
A future is a better abstraction than a callback in my opinion but that is subjective. We could keep using callbacks as well. The end result in the same. The new Controller interface uses Futures as well so standardizing makes sense here.

Regarding the current APIs, join group and sync group are definitely relying on callbacks. For the others, I don’t remember from the top of my head. In the new group coordinator, all of them will be executed in a different thread so they must be async.

Reply:
I don't have a strong opinion on callback vs future (maybe a slight bias for me towards callbacks because I'm more familiar, but not a good reason to change the design) but I do think it makes sense to be consistent. If we plan on using futures in the new controller interface then we can go ahead with using futures here too.